### PR TITLE
fix: dataloader keys manipulation logic when v2 vs v3

### PR DESCRIPTION
**Pull request by onhate**
- fixed logic for dataloader keys manipulation when client v2 vs v3 (v3 has `{S:''}` and v2 is `''`)
- run all tests both with v2 and v3

### CODE REVIEW

File: test/crypto.ts [hunk 1]
```diff
@@ -50,10 +50,18 @@ test('Get', async() => {
 
 test('Get raw ', async() => {
     user = await User.get({id: user.id}, {hidden: true, parse: false})
-    expect(user.pk.S).toMatch(/^User#/)
-    expect(user.email.S).toBeDefined()
-    expect(user.email.S).not.toMatch('peter@example.com')
-    expect(user.email.S).toMatch(/^primary/)
+    if (isV3()) {
```
Comment (mobsense) [inline test/crypto.ts:53]: User.get should not ever be returning email.S.  Should these values not be unmarshalled into just a pure email?

File: test/unique.ts [hunk 2]
```diff
@@ -39,8 +39,16 @@ test('Create user 1', async() => {
     let items = await table.scanItems()
     expect(items.length).toBe(3)
 
-    let unique = items.filter((item) => item.pk.S.indexOf('interpolated') >= 0)
-    let pk = unique[0].pk.S
+    let pk = (() => {
+        if (isV3()) {
+            let unique = items.filter((item) => (item.pk.S).indexOf('interpolated') >= 0)
+            return unique[0].pk.S
```
Comment (mobsense) [inline test/unique.ts:45]: Same as per previous comment. V3 and V2 should be returning identical data.

File: src/Table.js [hunk 3]
```diff
@@ -7,10 +7,10 @@
 import Crypto from 'crypto'
 import UUID from './UUID.js'
 import ULID from './ULID.js'
-import {Expression} from './Expression.js'
-import {Schema} from './Schema.js'
-import {Metrics} from './Metrics.js'
-import {OneTableError, OneTableArgError} from './Error.js'
+import { Expression } from './Expression.js'
```
Comment (mobsense) [inline src/Table.js:10]: Formatting.

**Comment by mobsense** [comment]
Thanks for the PR.

General question ... should not V2 and V3 be returning identical results. We don't want to be surfacing V2/V3 to the user (in general) and the loader should be returning identical results regardless of the SDK.

**Comment by onhate** [comment]
Hey, hi @mobsense, the loader is not returning different values, I just enabled a way to run all tests both on v2 and v3 and some unrelated tests with dataloader failed, all of them when `parse: false` which I would consider fair as v2 uses document client that already unmarshall data

**Comment by mobsense** [comment]
Can you help me understand the case with dataloader failing and needing parse: false?  In general, I've tried to avoid V2/V3 percolating up even in tests. Especially for V3 as that will become the dominant SDK and, again in general, we don't want to be surfacing the underlying DynamoDB encoding unmarshalled format.

**Comment by onhate** [comment]
The daloader does not needs parse as false, what happens is that as it uses the generated expressions to map the response from dynamodb and different clients generates different expressions and responses

**Comment by onhate** [comment]
If you check the dataloader test you will see the response is always parsed correctly both on v2 and v3, the difference on tests from v2 and v3 only appears when asking for unparsed values, which is correct in guess

**Comment by onhate** [comment]
@mobsense here is the flow and why loader needs to deal with ddb client specifics:

1. model.load is called
2. data loader accumulates the expressions to batch get
3. expressions can be either v2 (unmarshalled) or v3 (marshalled)
4. batch get does not supports duplicate keys on batch get so we need to "cleanup" the array of expressions
5. as expressions can be either v2 or v3 we need (internally only) check if what is the version used to make the expressions unique https://github.com/sensedeep/dynamodb-onetable/pull/270/files#diff-a44f85a07e6bd8ef01c1de04a73f6db1da817c9bb48d523071f0f0f63ac20120R768
6. after batch get is run we need now to return an array of items matching the exact same position they were requested

for example: 

if you requested `[{pk:1, sk:1}, {pk:1,:sk1},{pk2:sk:2}]` I run a batch get only for `[{pk:1, sk:1}, {pk2:sk:2}]` but the response required 3 items so I need to loop the responses (that can be either v2 or v3) and match the key with the item to return 3 items matching the exact same order they were requested.

In summary, internally we need to know if it is v2 or v3 to make keys unique and process response but the output of the dataloader is exactly the same as if it was a `get` call


Why differences for v2 and v3 on tests?

Well, I noticed we were not fully testing v2 so I made a small change on tests to run all suites both with v2 and v3 and when I did that some unrelated with dataloader tests, all using `parse: false` failed because they were only written for v3. In my opinion if you ask for unparsed values and we return raw values from dynamodb it's expected that the response will vary depending on the client you use.

File: test/crypto.ts (hunk 1 above)
Comment (onhate) [reply test/crypto.ts:53]: as explained here https://github.com/sensedeep/dynamodb-onetable/pull/270#issuecomment-1030811573 it is because this test asks for `parsed: false` values

File: test/unique.ts (hunk 2 above)
Comment (onhate) [reply test/unique.ts:45]: https://github.com/sensedeep/dynamodb-onetable/pull/270#discussion_r800164728

**Comment by mobsense** [comment]
Looks good. Just need the conflict with index.js resolved to merge.

**Comment by onhate** [comment]
fixed, thanks @mobsense

**Comment by mobsense** [comment]
Great, thank you!
